### ISSUE TITLE
Decide on client <-> hardware wallet API

### ISSUE THREAD

**Issue by hkjn** [hkjn/lnhw#9]
I'm filing a ticket to track our discussions to date about settling on an API between `client` and `hardware` layers.

Our current thinking is to initiate communication from the `client` layer, and use the serial port to send text-encoded requests to the `hardware` layer, which will return similarly text-encoded responses.

The responses might be delayed until the UI on the `hardware` side has gotten confirmation that the user wants to take a certain action.

**Comment by TheCharlatan** [comment]
@stephaniestroka , @stepansnigirev and me decided to use json to pass messages between client and hardware. The messages should not contain newlines, such that every newline can be cleanly read from the serial interface as a new message.

**Comment by hkjn** [comment]
Cool, sounds good to me.

So for the "first bite", I guess the request from `client` to `hardware` will look something like:

Request from `client`:
```
{
  "request": "get_hsm_secret"
}
```

Response from `hardware` in case of success:
```
{
  "status": "success",
  "payload": "abcdef12345[..]"  
}
```

And in case of failure:
```
{
  "status": "failure",
  "payload": "user did not approve in time"  
}
```

**Comment by stephaniestroka** [comment]
@TheCharlatan, here are the responses that are to be expected:
on success response, as above:
```{ "status": "success",  "payload": "abcdef1234abcdef1234abcdef1234ff"  }```
on failure:
- invalid JSON:
```
{  "status": "failure", "payload": "failed to parse input" }
```
- cancelled by user:
```
{  "status": "failure", "payload": "user cancelled" }
```
- invalid request value (no "get_hsm_secret" value):
```
{  "status": "failure", "payload": "invalid request" }
```
- missing "request" key:
```
{  "status": "failure", "payload": "missing request field" }
```

**Comment by hkjn** [comment]
I'll catch this issue up to what we ended up doing now that the hackathon is over..

What we ended up doing for the scope of the hackathon was:

1. get JSON serialization working on `hardware` side
2. (eventually, after much flailing) get JSON deserialization working on `client` side, using `yajl` library (https://github.com/lloyd/yajl/)
3. struggle to integrate the sample code that used `yajl` (https://github.com/hkjn/lightning/blob/master/hardwared/hardware.c) with the lightningd build system
4. run out of time (step 3 was late the night before the end of the hackathon) and instead quickly build a hack that uses a separate `rw` (https://github.com/hkjn/lightning/blob/master/readwrite.c) binary
that's called from the code paths that require user interaction (e.g https://github.com/hkjn/lightning/blob/master/hsmd/hsm.c#L676)

JSON is still a good idea for when we pick this up again!
